src/symmetric_crypto/metadata.rs

```diff
@@ -0,0 +1,141 @@
+use crate::CryptoCoreError;
+use serde::{Deserialize, Serialize};
+use std::convert::TryInto;
+
+/// Attempt getting the length of this slice as an u32 in 4 endian bytes and
+/// return an error if it overflows
+fn get_u32_len(slice: &[u8]) -> Result<[u8; 4], CryptoCoreError> {
+    u32::try_from(slice.len())
+        .map_err(|_| {
+            CryptoCoreError::InvalidSize(
+                "Slice of bytes is too big to fit in 2^32 bytes".to_string(),
+            )
+        })
+        .map(u32::to_be_bytes)
+}
+
+/// Scans a slice sequentially, updating the cursor position on the fly
+pub struct BytesScanner<'a> {
+    bytes: &'a [u8],
+    start: usize,
+}
+
+impl<'a> BytesScanner<'a> {
+    /// Return a new byte scanner object.
+    #[must_use]
+    pub const fn new(bytes: &'a [u8]) -> Self {
+        BytesScanner { bytes, start: 0 }
+    }
+
+    /// Returns a slice of the next `size` bytes or an error if less is
+    /// available
+    pub fn next(&mut self, size: usize) -> Result<&'a [u8], CryptoCoreError> {
+        let end = self.start + size;
+        if self.bytes.len() < end {
+            return Err(CryptoCoreError::InvalidSize(format!(
+                "{size}, only {} bytes available",
+                self.bytes.len() - self.start
+            )));
+        }
+        let chunk = &self.bytes[self.start..end];
+        self.start = end;
+        Ok(chunk)
+    }
+
+    /// Read the next 4 big endian bytes to return an u32
+    pub fn read_u32(&mut self) -> Result<u32, CryptoCoreError> {
+        Ok(u32::from_be_bytes(self.next(4)?.try_into().map_err(
+            |_e| CryptoCoreError::ConversionError("invalid u32".to_string()),
+        )?))
+    }
+
+    /// Returns the remainder of the slice
+    pub fn remainder(&mut self) -> Option<&'a [u8]> {
+        if self.start >= self.bytes.len() {
+            None
+        } else {
+            let remainder = &self.bytes[self.start..];
+            self.start = self.bytes.len();
+            Some(remainder)
+        }
+    }
+}
+
+/// Metadata encrypted as part of the header
+///
+/// The `uid` is a security parameter:
+///  - when using a stream cipher such as AES, it uniquely
+///    identifies a resource, such as a file, and is part of the AEAD of every
+///    block when symmetrically encrypting data. It prevents an attacker from
+///    moving blocks between resources.
+///  - when using FPE, it is the "tweak"
```

Comment:
```suggestion
///  - when using FPE, it is the "tweak" (see research paper)
```

btw it would be great if we can give a link of this paper here